File: ThreemaFramework/Utilities/PhoneNumberNormalizer.m
#import "PhoneNumberNormalizer.h"
#import "NBPhoneNumberUtil.h"
#import <CoreTelephony/CTTelephonyNetworkInfo.h>
#import <CoreTelephony/CTCarrier.h>
#import "NBMetadataHelper.h"

@interface PhoneNumberNormalizer ()

@property NBPhoneNumberUtil *phoneNumberUtil;

@end

@implementation PhoneNumberNormalizer

+ (PhoneNumberNormalizer*)sharedInstance {
    static PhoneNumberNormalizer *sharedInstance;
    static dispatch_once_t pred;
    dispatch_once(&pred, ^{
        sharedInstance = [[PhoneNumberNormalizer alloc] init];
        
        sharedInstance.phoneNumberUtil = [NBPhoneNumberUtil sharedInstance];
    });
    return sharedInstance;
}

- (NSString*)phoneNumberToE164:(NSString*)phoneNumber withDefaultRegion:(NSString*)defaultRegion prettyFormat:(NSString**)prettyFormat {
    
    NSError *error = nil;
    NBPhoneNumber *parsed = [_phoneNumberUtil parse:phoneNumber defaultRegion:defaultRegion error:&error];
    if (parsed == nil)
        return nil;
    
    NSString *e164WithPlus = [_phoneNumberUtil format:parsed numberFormat:NBEPhoneNumberFormatE164 error:&error];
    if (e164WithPlus == nil)
        return nil;
    
    NSString *result = [e164WithPlus stringByTrimmingCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@"+"]];
    
    if (prettyFormat != nil)
        *prettyFormat = [_phoneNumberUtil format:parsed numberFormat:NBEPhoneNumberFormatINTERNATIONAL error:&error];
    
    return result;
}

- (NSString *)examplePhoneNumberForRegion:(NSString *)region {
    
    NSError *error = nil;
    
    NBPhoneNumber *exampleNumber = [_phoneNumberUtil getExampleNumber:region error:&error];
    if (exampleNumber == nil)
        return nil;
    
    return [_phoneNumberUtil format:exampleNumber numberFormat:NBEPhoneNumberFormatNATIONAL error:&error];
}

- (NSString *)exampleRegionalPhoneNumberForRegion:(NSString *)region {
    
    NSError *error = nil;
    
    NBPhoneNumber *exampleNumber = [_phoneNumberUtil getExampleNumber:region error:&error];
    if (exampleNumber == nil) {
        return nil;
    }
    
    NSString *number = [_phoneNumberUtil format:exampleNumber numberFormat:NBEPhoneNumberFormatINTERNATIONAL error:&error];

    return [self regionalPartForPhoneNumber:number];
}

- (NSString *)regionalPartForPhoneNumber:(NSString *)phoneNumber {
    NSString *region = [self regionForPhoneNumber:phoneNumber];
    if (region) {
        NBMetadataHelper *metadataHelper = [[NBMetadataHelper alloc] init];
        NSString *code = [[metadataHelper countryCodeToCountryNumberDictionary] objectForKey:region];
    
        NSRange codeRange = [phoneNumber rangeOfString:code];
        if (codeRange.location != NSNotFound) {
            return [phoneNumber substringFromIndex:codeRange.location + codeRange.length];
        }
    }
    
    return phoneNumber;
}

- (NSString *)regionForPhoneNumber:(NSString *)phoneNumber {
    NSError *error = nil;
    
    NSString *defaultRegion = [PhoneNumberNormalizer userRegion];
    NBPhoneNumber *parsed = [_phoneNumberUtil parse:phoneNumber defaultRegion:defaultRegion error:&error];
    if (parsed == nil) {
        return nil;
    }
    
    return [_phoneNumberUtil getRegionCodeForNumber:parsed];
}

+ (NSString*)userRegion {
    /* try CTCarrier first as it's probably more reliable (who knows what the user's country setting may be) */
    CTTelephonyNetworkInfo *netInfo = [[CTTelephonyNetworkInfo alloc] init];
    NSDictionary<NSString *, CTCarrier *> *carriers = [netInfo serviceSubscriberCellularProviders];
    for (CTCarrier *carrier in carriers.allValues) {
        // Since iOS 16 is this function deprecated. It will return always "--".
        if (carrier.isoCountryCode != nil && ![carrier.isoCountryCode isEqualToString:@"--"]) {
            return [carrier.isoCountryCode uppercaseString];
        }
    }
    
    /* fall back to NSLocale */
    return [[NSLocale currentLocale] objectForKey:NSLocaleCountryCode];
}

@end

File: ThreemaFramework/Utilities/PhoneNumberNormalizer/PhoneNumberNormalizer.swift
import Foundation
import libPhoneNumber
import ObjectiveC

@objcMembers public final class PhoneNumberNormalizer: NSObject, PhoneNumberNormalizerProtocol {

    // MARK: - Private properties

    private let phoneNumberUtil: NBPhoneNumberUtil

    // MARK: - Lifecycle

    override public init() {
        self.phoneNumberUtil = NBPhoneNumberUtil.sharedInstance()
        super.init()
    }

    // MARK: - Public methods

    public func e164Format(from phoneNumber: String, defaultRegion: String) -> String? {
        guard
            let parsed = try? phoneNumberUtil.parse(phoneNumber, defaultRegion: defaultRegion),
            let formatted = try? phoneNumberUtil.format(parsed, numberFormat: .E164)
        else {
            return nil
        }

        return formatted.trimmingCharacters(in: CharacterSet(charactersIn: "+"))
    }

    public func prettyFormat(from phoneNumber: String, defaultRegion: String) -> String? {
        guard
            let parsed = try? phoneNumberUtil.parse(phoneNumber, defaultRegion: defaultRegion)
        else {
            return nil
        }

        return try? phoneNumberUtil.format(parsed, numberFormat: .INTERNATIONAL)
    }

    public func examplePhoneNumber(for region: String) -> String? {
        guard
            let exampleNumber = try? phoneNumberUtil.getExampleNumber(region)
        else {
            return nil
        }

        return try? phoneNumberUtil.format(exampleNumber, numberFormat: .NATIONAL)
    }

    public func exampleRegionalPhoneNumber(for region: String) -> String? {
        guard
            let exampleNumber = try? phoneNumberUtil.getExampleNumber(region),
            let number = try? phoneNumberUtil.format(exampleNumber, numberFormat: .INTERNATIONAL)
        else {
            return nil
        }

        return regionalPart(from: number)
    }

    public func regionalPart(from phoneNumber: String) -> String? {
        guard
            let metadataHelper = NBMetadataHelper(),
            let region = region(from: phoneNumber)
        else {
            return nil
        }

        if
            let code = metadataHelper.countryCodeToCountryNumberDictionary()[region] as? String,
            let codeRange = phoneNumber.range(of: code) {
            return String(phoneNumber[codeRange.upperBound...]).trimmingCharacters(in: .whitespaces)
        }

        return nil
    }

    public func userRegion() -> String {
        Locale.current.region?.identifier ?? ""
    }

    public func region(from phoneNumber: String) -> String? {
        guard let parsed = try? phoneNumberUtil.parse(phoneNumber, defaultRegion: nil) else {
            return nil
        }
        return phoneNumberUtil.getRegionCode(for: parsed)
    }
}
